### PR TITLE
Force successor permissions

### PR DESCRIPTION
**Pull request by 13steinj**
When a top mod removes him/herself, there was no insurance to the fact that the next mod should recieve full permissions. This commit makes a common /r/redditrequest automated and should reduce strain on that subreddit. This also automates the removal of spam/deleted users if they would be the successor.

I've made the two properties in subreddit.py just because there are other semi-issues that rely on what would be provided by those properties, so for the sake of making it easier on future issues I was going to fix, I made them properties instead of determining them in api.py

### CODE REVIEW

**Comment by umbrae** [comment]
:eyeglasses: @Deimos

File: r2/r2/models/subreddit.py [hunk 1]
```diff
@@ -553,6 +553,20 @@ def remove_moderator(self, user, **kwargs):
     def moderators(self):
         return self.moderator_ids()
 
+    @property
+    def top_moderator_account(self):
+        if len(self.moderators) >= 1:
+            return Account._byID(self.moderators[0])
+        else:
+            return FakeAccount()
```
Comment (13steinj) [inline r2/r2/models/subreddit.py]: Just realized you guts may want this line to actually return `None` instead of a `FakeAccount()`

File: r2/r2/models/subreddit.py [hunk 2]
```diff
@@ -553,6 +553,20 @@ def remove_moderator(self, user, **kwargs):
...
+
+    @property
+    def top_moderator(self):
+        if len(self.moderators) >= 1:
+            return self.get_moderator(self.top_moderator_account)
+        else:
+            return FakeSRMember(ModeratorPermissionSet)
```
Comment (13steinj) [inline r2/r2/models/subreddit.py]: Same here

File: r2/r2/controllers/api.py [hunk 3]
```diff
@@ -795,6 +800,86 @@ def check_api_friend_oauth_scope(self, type_):
...
+                potentialsuccessor = Account._byID(user_id)
+                potentialsuccessor_as_mod = sr.get_moderator_invite(potentialsuccessor)
+                potentialsuccessor_perms = potentialsuccessor_as_mod.get_permissions()
+                # uninvite each next successor if said successor is deleted / marked as spam
+                if potentialsuccessor._deleted or potentialsuccessor._spam:
+                    if potentialsuccessor._deleted:
+                        badsuccessordescription = "user is [deleted]"
+                    elif potentialsuccessor._spam:
+                        badsuccessordescription = "user is shadowbanned"
+                    sr.remove_moderator(potentialsuccessor)
```
Comment (13steinj) [inline r2/r2/controllers/api.py]: Whoops. Should be `sr.remove_moderator_invite`

File: r2/r2/controllers/api.py [hunk 4]
```diff
@@ -795,6 +800,86 @@ def check_api_friend_oauth_scope(self, type_):
...
+                        sr,
+                        c.user,
+                        'uninvitemoderator',
+                        target=potentialsuccessor,
+                        details='account is not a valid successor',
+                        description=badsuccessordescription
+                    )
+                    continue
+                # if the successor does not have all perms, give it to him/her
+                if not potentialsuccessor_perms['all']:
```
Comment (13steinj) [inline r2/r2/controllers/api.py]: I didn't use `has_permissions` here nor on the mod cycle because I couldn't get it to work here. Weirdly enough this syntax worked fine.

File: r2/r2/controllers/api.py [hunk 5]
```diff
@@ -795,6 +800,86 @@ def check_api_friend_oauth_scope(self, type_):
                 # this request.
                 self._auth_error(403, "insufficient_scope")
 
+    def _handle_mod_successorship(self, sr):
```
Comment (13steinj) [inline r2/r2/controllers/api.py]: You may want these cycles being just one, the list to be cycled through determined via `getattr`

File: r2/r2/controllers/api.py [hunk 6]
```diff
@@ -870,6 +931,11 @@ def POST_unfriend(self, nuser, iuser, container, type):
         if type in ("friend", "enemy") and container != c.user:
             abort(403, 'forbidden')
         fn = getattr(container, 'remove_' + type)
+
+        # if the top mod is removing themself
```
Comment (13steinj) [inline r2/r2/controllers/api.py]: This comment is slightly inaccurate: should be "if the top mod is removing themself or an admin is removing them"

File: r2/r2/controllers/api.py [hunk 7]
```diff
@@ -752,6 +754,9 @@ def POST_leavecontributor(self, container):
         """
         if container and container.is_contributor(c.user):
             container.remove_contributor(c.user)
+            if container.is_moderator_with_perms(c.user, 'access'):
```
Comment (Deimos) [inline r2/r2/controllers/api.py]: This seems unrelated, probably shouldn't be in here (or at least in a separate commit). I'm also not really sure why it's being added, why do we want a mod log entry when mods remove themselves as a contributor, when there isn't one for any other user removing themselves?

**Comment by Deimos** [comment]
I think this change ended up being way over-complicated. There's no need to have moderators being auto-removed as part of this process, and I think the situation where the only moderator steps down while having a pending invite out with non-full permissions is probably so rare that it's really not worth worrying about.

Can we simplify it down to do *only* this: "if a top moderator leaves (or is removed), give full permissions to the first non-deleted moderator below them on the list"? It should also be done in the `remove_moderator()` function inside the `Subreddit` class, not in the individual API endpoints.

:nail_care:

File: r2/r2/controllers/api.py (hunk 7 above)
Comment (13steinj) [inline r2/r2/controllers/api.py]: @Deimos The reasoning I had was that when mods remove anyone or themselves from being a mod, a mod action was created; and there's also a mod action made when removing themselves as a contributor if they remove themselves from the /about/contributor page (as that itself is done via `POST_unfriend` and not `POST_leavecontributor`)

**Comment by 13steinj** [comment]
@Deimos :haircut: 

I didn't remove the remove_self just in case, and left a comment [here](https://github.com/reddit/reddit/pull/1420#discussion_r44602352). If needed I can remove it (and or rebase from master to resolve merge conflicts)

File: r2/r2/controllers/api.py (hunk 7 above)
Comment (Deimos) [inline r2/r2/controllers/api.py]: Alright. I don't think it's really a big deal, but either way it needs to be in a separate commit since it's completely unrelated to the "automatic permissions to successor" stuff.

File: r2/r2/controllers/api.py [hunk 8]
```diff
@@ -870,6 +873,7 @@ def POST_unfriend(self, nuser, iuser, container, type):
         if type in ("friend", "enemy") and container != c.user:
             abort(403, 'forbidden')
         fn = getattr(container, 'remove_' + type)
+
```
Comment (Deimos) [inline r2/r2/controllers/api.py]: Nit: please remove the newline being added here. We try to avoid making whitespace changes except when we're already adjusting code for another reason, and this is separated from the actual changes.

File: r2/r2/controllers/api.py [hunk 9]
```diff
@@ -883,7 +887,16 @@ def POST_unfriend(self, nuser, iuser, container, type):
                 contributor='removecontributor',
                 muted='unmuteuser',
             ).get(type, None)
-            ModAction.create(container, c.user, action, target=victim)
+            details = None
+            if (c.user == victim and
```
Comment (Deimos) [inline r2/r2/controllers/api.py]: This `if` is extremely difficult to follow (and `is` shouldn't be used to check string equality).

I really don't think it's even necessary, I'd probably just do `if c.user == victim and c.site.is_moderator(c.user):` and set the `remove_self` note on that. I can't see any way that the distinction of "moderator that has permissions to remove others is removing themselves" vs. "user is a moderator that doesn't have permissions to remove others, but is leaving this status" is important at all.

File: r2/r2/models/subreddit.py [hunk 10]
```diff
@@ -553,11 +580,25 @@ def remove_moderator(self, user, **kwargs):
...
+        if len(self.moderators) > 0:
+            return self.get_moderator(self.top_moderator_account)
+
     def moderators_with_perms(self):
         return collections.OrderedDict(
             (r._thing2_id, r.get_permissions())
             for r in self.each_moderator())
 
+    @property
+    def moderator_invites(self):
```
Comment (Deimos) [inline r2/r2/models/subreddit.py]: This property isn't used any more, and can be removed.

File: r2/r2/models/subreddit.py [hunk 11]
```diff
@@ -553,11 +580,25 @@ def remove_moderator(self, user, **kwargs):
     def moderators(self):
         return self.moderator_ids()
 
+    @property
+    def top_moderator_account(self):
```
Comment (Deimos) [inline r2/r2/models/subreddit.py]: Let's just drop both of these properties too, I think. We can just check the IDs directly above, so we don't really need `top_moderator_account` or `top_moderator`.

File: r2/r2/models/subreddit.py [hunk 12]
```diff
@@ -540,6 +540,33 @@ def remove_moderator(self, user, **kwargs):
         hook = hooks.get_hook("subreddit.remove_moderator")
         hook.call(subreddit=self, user=user)
 
+        if user == self.top_moderator_account and len(self.moderators) > 1:
```
Comment (Deimos) [inline r2/r2/models/subreddit.py]: This gets into some pretty heavy indenting, I think this is going to be a lot cleaner as a separate function. I'd just do something like this:

    # if the top mod is being removed, make sure the new one has full perms
    if user._id == self.moderators[0]:
        self._handle_mod_succession()

(There's no need to check that there's more than 1 moderator, that will be automatically handled by the list of potential successors being empty)

Then move the code inside here to a new function, which will save some indentation and make it a little easier overall.

File: r2/r2/models/subreddit.py [hunk 13]
```diff
@@ -540,6 +540,33 @@ def remove_moderator(self, user, **kwargs):
...
+            from r2.models import ModAction
+            fullperms = self.moderator_permission_class.loads('+all')
+            # cycle through each available moderator
+            for user_id in self.moderators[1:]:
```
Comment (Deimos) [inline r2/r2/models/subreddit.py]: It might be a bit easier to use `for user_id, perms in self.moderators_with_perms():` here, then you don't have to get the rel/perms manually each time.

Comment (Deimos) [inline r2/r2/models/subreddit.py]: Note that to make this work most easily, you should move the call to this new successor function **below** the point in `remove_moderator()` where `ret = super(Subreddit, self).remove_moderator(user, **kwargs)` is called (so that the user being removed is no longer a mod, and won't be considered as a candidate, and you don't have to worry about skipping them or anything).

File: r2/r2/models/subreddit.py [hunk 14]
```diff
@@ -540,6 +540,33 @@ def remove_moderator(self, user, **kwargs):
...
+                successor = Account._byID(user_id)
+                # skip deleted / shadowbanned users
+                if successor._deleted or successor._spam:
```
Comment (Deimos) [inline r2/r2/models/subreddit.py]: I think we had probably better not skip `._spam` users for now. We still more or less consider them "valid mods" for most things (they're still in the mod list and such), so if we skip over them there's a risk of a weird case where a temporarily-shadowbanned user gets skipped as a successor, and full perms actually pass to the third (or lower) mod instead of the second. I think just skipping deleted ones is best to avoid that case for now.

File: r2/r2/models/subreddit.py [hunk 15]
```diff
@@ -540,6 +540,33 @@ def remove_moderator(self, user, **kwargs):
...
+                    continue
+                successor_rel = self.get_moderator(successor)
+                successor_perms = successor_rel.get_permissions()
+                # if the successor does not have all perms, give it to him/her
+                if not successor_perms['all']:
```
Comment (Deimos) [inline r2/r2/models/subreddit.py]: Just to make this a little simpler (and reduce indentation more), when you're doing this in the new function you can just do something more like:

    # if the successor already has full perms, we don't need to do anything
    if successor_perms["all"]:
        return

And then the code for setting permissions and creating the mod log entry doesn't need to be inside an `if:`.

**Comment by Deimos** [comment]
Mostly just organization suggestions, looks pretty good overall. :nail_care:

**Comment by 13steinj** [comment]
@Deimos :haircut: should I rebase to resolve merge conflicts as well as squash the extra commits together?

File: r2/r2/models/subreddit.py [hunk 16]
```diff
@@ -536,12 +536,46 @@ def add_moderator(self, user, **kwargs):
 
         return super(Subreddit, self).add_moderator(user, **kwargs)
 
+    def _handle_mod_succession(self):
+        if len(self.moderators) == 0:
+            return
```
Comment (13steinj) [inline r2/r2/models/subreddit.py]: This check isn't actually needed, as the for loop would just do nothing instead...but from something I asked a month or so ago on stackoverflow, iterating over an empty list would still take up some processing power and an extremely small amount of time, so I added it anyway.

**Comment by 13steinj** [comment]
@Deimos :haircut: I've re-updated since I viewed this earlier today and noticed some things were inefficient.

**Comment by bsimpson63** [comment]
Sorry, but this repository is no longer accepting pull requests. https://www.reddit.com/r/changelog/comments/6xfyfg/an_update_on_the_state_of_the_redditreddit_and/
